File: 2020/day3/part2/main.go
package part2

import (
	"advent-of-code/advent_of_code/day3/part1"
	"advent-of-code/advent_of_code/helpers"
	"fmt"
)

func Slope(rows []string, slide int, lineSkip int) int {

	count := 0
	index := 0
	lineNumber := 0

	// go's while loops are weird
	for lineNumber < len(rows) {

		row := rows[lineNumber]
		res := part1.Loop(len(row), index, row)
		if res {
			count++
		}
		index += slide
		lineNumber += lineSkip

	}

	return count

}

func Main(filepath string) int {
	inputFile, err := helpers.ReadFileLineByLine(filepath)
	if err != nil {
		fmt.Println("%n", err)
	}

	conditions := [][]int{
		{1, 1}, {1, 3}, {1, 5}, {1, 7}, {2, 1},
	}
	var counts []int

	for _, list := range conditions {
		counts = append(counts, Slope(inputFile, list[1], list[0]))
	}

	// go doesn't have generics, much sad
	returnValue := 1
	for _, val := range counts {
		returnValue *= val
	}

	return returnValue
}

File: 2020/day3/part2/main.py
from functools import reduce
from typing import List, Tuple


from advent_of_code.helpers.helpers import read_file_line_by_line
from advent_of_code.day3.part1.main import loop


def slope(rows: List[str], slide: int, line_skip: int = 1) -> int:

    count: int = 0
    index: int = 0
    line_number: int = 0

    while line_number < len(rows):
        row: List[str] = rows[line_number]
        res: int = loop(len(row), index=index, row=row)
        count += 1 if res else 0
        index += slide
        line_number += line_skip

    return count


def main(file: str = "") -> int:

    input_file: List[str] = read_file_line_by_line(filepath=file)

    conditions: List[Tuple[int, int]] = [
        (1, 1), (1, 3), (1, 5), (1, 7), (2, 1)
    ]
    counts: List[int] = []

    for down, right in conditions:
        counts.append(
            slope(rows=input_file, slide=right, line_skip=down)  
        )

    return reduce((lambda x, y: x * y), counts)


